R/validate_inputs.R

```diff
@@ -31,68 +31,45 @@ do_validate_shape <- function(shape) {
   # and it's only 2.5MB large).  A caching layer will help, but this
   # is going to lock things up enough we might need to do it
   # asynchronously.
-  json <- geojsonio::geojson_read(shape, method = "local")
-  assert_single_country(json)
+  json <- hintr_geojson_read(shape)
+  assert_single_country(json, "shape")
   assert_area_id_exists(json)
   # Then we have to *reread* the file now that we know that it is
   # valid, but but this is not too slow, especially as the file is now
   # in cache (but still ~1/20s)
   json_verbatim(read_string(shape))
 }
 
-assert_single_country <- function(json) {
-  country <- vapply(json$features, function(x) {
-    x$properties$iso3
-  }, character(1))
-  if (length(unique(country)) != 1) {
-    stop(sprintf(
-      "Shape file contains regions for more than one country. Got countries %s.",
-      toString(unique(country))))
-  }
-  invisible(TRUE)
-}
-
-assert_area_id_exists <- function(json) {
-  contains_id <- vapply(json$features, function(x) {
-    !is_empty(x$properties$area_id)
-  }, logical(1))
-  if (!all(contains_id)) {
-    missing_count <- sum(!contains_id)
-    stop(
-      sprintf(
-        "Shape file does not contain an area ID for each region. Missing ID for %s %s.",
-        missing_count,
-        ngettext(missing_count, "feature", "features")
-      )
-    )
-  }
-  invisible(TRUE)
-}
-
 #' Validate population file.
 #'
 #' Check that population file can be read.
 #'
-#' @param shape Path to input population file.
+#' @param population Path to input population file.
 #'
 #' @return An error if invalid.
 #' @keywords internal
 do_validate_population <- function(population) {
   population <- read_csv(population, header = TRUE)
-  ## Perhaps this kind of assertion should just be checked by json schema?
+  assert_single_country(population, "population")
```

Comment:
This could be a nice place to use the defer package?

Reply:
I think I will leave this for a separate PR once we get that package on drat.